### ISSUE TITLE
Is there support to detect if any device was plugged/unplugged?

### ISSUE THREAD

**Issue by rigorio** [oshi/oshi#715]
I need my hardware/software info to be always updated, so I'm planning to update it every time a device is plugged/unplugged. Is there anyway in oshi to detect this?

**Comment by dbwiddis** [comment]
OSHI doesn't do any real time monitoring of events/callbacks.    It does list USB devices and you could possibly compare old vs. new lists to detect differences there.

I'm considering a similar functionality for the entire device tree (rather than just USB) but don't expect that anytime soon unless someone else wants to do it!

**Comment by shanmukhateja** [comment]
I think that is a great idea. On the Linux front, we have DBus that can listen to custom events for us. I certainly know there's Python bindings to DBus however I'm not sure on the Java end.

> I'm considering a similar functionality for the entire device tree (rather than just USB)

That is indeed interesting. Could you elaborate a little?

**Comment by dbwiddis** [comment]
> That is indeed interesting. Could you elaborate a little?

Here's a look at the [Windows Device Tree](https://docs.microsoft.com/en-us/windows-hardware/drivers/gettingstarted/device-nodes-and-device-stacks).  We currently use a portion of this tree to build the Windows USB device tree in OSHI.  See the `putChildrenInDeviceTree()` method in `WindowsUSBDevice.java`.  Now imagine instead of starting with the USB Controllers, we start all the way up at the root node.  (See the [test case I submitted to the JNA project](https://github.com/java-native-access/jna/blob/c7419aff496080278f9d0ff066eeda377cca43ae/contrib/platform/test/com/sun/jna/platform/win32/Cfgmgr32Test.java#L45) that does this for the first layer; apply recursion as required.)  

macOS has the IO Registry, which we already explore for the USB tree; like Windows we'd just start higher. Linux has the device tree in the sysfs: `/sys/firmware/devicetree/base` can be navigated.  I'm guessing our current `udev` parsing from the USB code can also work there. Solaris has a `/devices` directory or the `prtconf` command.  FreeBSD has `sys/boot/fdt/dts`.

Basically the tree navigating structure is already there, we'd just go up a higher level so instead of only finding USB devices, we'd find ALL devices and display/list them.  We could then compare devices between successive runs and identify differences.

**Comment by rigorio** [comment]
Sorry I just got back from vacation.

> OSHI doesn't do any real time monitoring of events/callbacks. It does list USB devices and you could possibly compare old vs. new lists to detect differences there.

I see. So would you suggest for now something like for every 10 seconds, retrieve all USB devices and compare them to the currently saved USB devices?

> I'm considering a similar functionality for the entire device tree (rather than just USB) but don't expect that anytime soon unless someone else wants to do it!

This is really interesting! Hopefully when I gain enough experience in Java, I'll be able to do it :smile: 
Though that might be after a year or so since I've only been involved in Java dev for barely a year. So I do hope someone else is able to do it earlier.

**Comment by dbwiddis** [comment]
While I may have longevity on you, I'm not a professional dev.  I got into this project to fix a bug and kinda stuck around.  With a year of dev experience you're probably on par if not better than me. :)

And really, most of the code needed for a device tree type solution is already there with the existing USB code on the various platforms, you just change the starting point/root node.

As far as a polling plan... 10 seconds is probably aggressive for something that doesn't change often (like pluggable devices).  OSHI wasn't designed as a real time monitoring system, though, more of a snapshot/state provider.

**Comment by rigorio** [comment]
> While I may have longevity on you, I'm not a professional dev. I got into this project to fix a bug and kinda stuck around. With a year of dev experience you're probably on par if not better than me. :)
> 
> And really, most of the code needed for a device tree type solution is already there with the existing USB code on the various platforms, you just change the starting point/root node.


Well I'll try to study Oshi source code on my free time. But really, I still have a long ways to go :smile: 

> As far as a polling plan... 10 seconds is probably aggressive for something that doesn't change often (like pluggable devices). OSHI wasn't designed as a real time monitoring system, though, more of a snapshot/state provider.

Got it. It's more clear for me now. Thank you so much!! :)
